README.md
[<p align="center"><img src="mas-cli.png" alt="mas-cli" width="450" height="auto"></p>][mas-cli]

# mas-cli

A simple command line interface for the Mac App Store. Designed for scripting and automation.

[![Software License](https://img.shields.io/badge/license-MIT-lightgrey.svg)](https://github.com/mas-cli/mas/blob/master/LICENSE)
[![Swift 5](https://img.shields.io/badge/Language-Swift_5-orange.svg)](https://swift.org)
[![GitHub Release](https://img.shields.io/github/release/mas-cli/mas.svg)](https://github.com/mas-cli/mas/releases)
[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)
[![Pipeline Status](https://jenkins.log-g.co/buildStatus/icon?job=mas-cli/mas/master)](https://jenkins.log-g.co/job/mas-cli/job/mas/job/master/)

## 📲 Install

### 🍺 Homebrew

[Homebrew](http://brew.sh) is the preferred way to install:

```bash
brew install mas
```

### MacPorts

[MacPorts](https://www.macports.org/install.php) works as well:

```bash
sudo port install mas
```

⚠️ Note that macOS 10.14 (Mojave) is required to install mas from MacPorts or the core Homebrew formula.

### ☎️ Older macOS Versions

We provide a [custom Homebrew tap](https://github.com/mas-cli/homebrew-tap) with pre-built bottles
for all macOS versions since 10.11.

To install mas from our tap:

```bash
brew install mas-cli/tap/mas
```

#### Known Installation Issues On macOS Before 10.14.4

Starting with [mas 1.7.0](https://github.com/mas-cli/mas/releases/tag/v1.7.0), if you get
an error similar to `dyld: Symbol not found: _$s11SubSequenceSlTl` when running mas,
you'll need to install the [Swift 5 Runtime Support for Command Line Tools](https://support.apple.com/kb/DL1998).

Alternatively, you can:

- Update to macOS 10.14.4 or later
- Install Xcode 10.2 or later at `/Applications/Xcode.app`
- Rebuild mas from source using Xcode 11.4 or later

### 🐙 GitHub Releases

Alternatively, binaries are available in the [GitHub Releases](https://github.com/mas-cli/mas/releases)

## 🤳🏻 Usage

Each application in the Mac App Store has a product identifier which is also
used for mas-cli commands. Using `mas list` will show all installed
applications and their product identifiers.

```bash
$ mas list
446107677 Screens
407963104 Pixelmator
497799835 Xcode
```

It is possible to search for applications by name using `mas search` which
will search the Mac App Store and return matching identifiers.
Include the `--price` flag to include prices in the result.

```bash
$ mas search Xcode
497799835 Xcode
688199928 Docs for Xcode
449589707 Dash 3 - API Docs & Snippets. Integrates with Xcode, Alfred, TextWrangler and many more.
[...]
```

To install or update an application simply run `mas install` with an
application identifier:

```bash
$ mas install 808809998
==> Downloading PaintCode 2
==> Installed PaintCode 2
```

If you want to install the first result that the `search` command returns, use the `lucky` command.

```bash
$ mas lucky twitter
==> Downloading Twitter
==> Installed Twitter
```

> Please note that this command will not allow you to install (or even purchase) an app for the first time:
use the `purchase` command in that case.

```bash
$ mas purchase 768053424
==> Downloading Gapplin
==> Installed Gapplin
```

> Please note that you may have to re-authenticate yourself in the App Store to complete the purchase.
This is the case if the application is not free or if you configured your account not to remember the credentials for free purchases.

Use `mas outdated` to list all applications with pending updates.

```bash
$ mas outdated
497799835 Xcode (7.0)
446107677 Screens VNC - Access Your Computer From Anywhere (3.6.7)
```

> `mas` is only able to install/update applications that are listed in the Mac App Store itself.
Use [`softwareupdate(8)`] utility for downloading system updates (like iTunes, Xcode Command Line Tools, etc)

To install all pending updates run `mas upgrade`.

```bash
$ mas upgrade
Upgrading 2 outdated applications:
Xcode (7.0), Screens VNC - Access Your Computer From Anywhere (3.6.7)
==> Downloading Xcode
==> Installed Xcode
==> Downloading iFlicks
==> Installed iFlicks
```

Updates can be performed selectively by providing the app identifier(s) to
`mas upgrade`

```bash
$ mas upgrade 715768417
Upgrading 1 outdated application:
Xcode (8.0)
==> Downloading Xcode
==> Installed Xcode
```

### 🚏📥 Sign-in

To sign into the Mac App Store for the first time run `mas signin`.

```bash
$ mas signin mas@example.com
==> Signing in to Apple ID: mas@example.com
Password:
```

> ⚠️ Due to breaking changes in the underlying API that mas uses to interact with the Mac App Store,
> the `signin` command has been temporarily disabled on macOS 10.13+ ⛔.
> For more information on this issue, see [#164](https://github.com/mas-cli/mas/issues/164).

If you experience issues signing in this way, you can ask to signin using a graphical dialog (provided by Mac App Store application):

```bash
$ mas signin --dialog mas@example.com
==> Signing in to Apple ID: mas@example.com
```

You can also embed your password in the command.

```bash
$ mas signin mas@example.com 'ZdkM4f$gzF;gX3ABXNLf8KcCt.x.np'
==> Signing in to Apple ID: mas@example.com
```

Use `mas signout` to sign out from the Mac App Store.

## 🍺 Homebrew integration

`mas` is integrated with [homebrew-bundle]. If `mas` is installed, and you run `brew bundle dump`,
then your Mac App Store apps will be included in the Brewfile created. See the [homebrew-bundle]
docs for more details.

## 💥 When something doesn't work

If you see the error "This redownload is not available for this Apple ID either because it was bought by a different user of the item was refunded or cancelled.", it's probably because you haven't installed the app through the App Store yet. See [#46](https://github.com/mas-cli/mas/issues/46#issuecomment-248581233).

If `mas` doesn't work for you as expected (e.g. you can't update/download apps), run `mas reset` and try again. If the issue persists, please [file a bug](https://github.com/mas-cli/mas/issues/new)! All your feedback is much appreciated ✨

## 📺 Using `tmux`

`mas` operates via the same system services as the Mac App Store. These exist as
separate processes with communication through XPC. As a result of this, `mas`
experiences similar problems as the pasteboard when running inside `tmux`. A
[wrapper tool exists](https://github.com/ChrisJohnsen/tmux-MacOSX-pasteboard) to
fix pasteboard behaviour which also works for `mas`.

You should consider configuring `tmux` to use the wrapper but if you do not wish
to do this it can be used on a one-off basis as follows:

```bash
brew install reattach-to-user-namespace
reattach-to-user-namespace mas install
```

## ℹ️ Build from source

You can now build from Xcode by opening `mas-cli.xcodeproj`, or from the Terminal:

```bash
script/build
```

Build output can be found in the `build/` directory within the project.

## ✅ Tests

The tests in this project are a recent work-in-progress.
Since Xcode does not officially support tests for command-line tool targets,
all logic is part of the MasKit target with tests in MasKitTests.
Tests are written using [Quick].

## 📄 License

mas-cli was created by [@argon](https://github.com/argon).
Code is under the [MIT license](LICENSE).

[homebrew-bundle]: https://github.com/Homebrew/homebrew-bundle
[mas-cli]: https://github.com/mas-cli/mas
[`softwareupdate(8)`]: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man8/softwareupdate.8.html
[Quick]: https://github.com/Quick/Quick


CHANGELOG.md
# mas Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

- 🍼 Formula updates from 1.8 release #319

## [v1.8.0] 💪🏼 arm64 support for M1 Macs - 2021-02-12

- ✨ Support for Apple Silicon (arm64) #310  
  thanks, [@chris-araman](https://github.com/chris-araman)!
  - ⚒️ Xcode 12.2 and macOS 10.15.4 required to build
- 🕊 Swift 5.2 #310
  - ⚒️ Xcode 11.4 and macOS 10.15.2 required to build
- 🐛 Specify desired `LD_RUNPATH_SEARCH_PATHS` at link-time #310
- 🐛 Default PREFIX to /opt/homebrew on Apple Silicon (arm64)
- ⬆️ Commandant (0.18.0) #310
- ⬆️ Nimble (9.0.0) #310
- ⬆️ Quick (3.1.0) #310
- 🎨 Improve `mas list` command output #278  
  thanks, [@danielbayley](https://github.com/danielbayley)!
- 🐛 Fix alignment of 2nd column of list command output #298
- ✨ `Makefile` #277

## [v1.7.1] 🐛 Search JSON Fix - 2020-09-06

- 🐛 Account for missing JSON fields in search results #299

## [v1.7.0] 🛍 Purchase Command - 2020-05-24

- ✨ New `purchase` command for purchasing free apps #264 (resolves #2, #145)  
  thanks, [@blochberger](https://github.com/blochberger)!
- 🐟 Seriously more interactive fish completions #242  
  thanks, [@lwolfsonkin](https://github.com/lwolfsonkin)!
  🍺Formula 1.7 #275
- ⬆️ Nimble (8.0.9) #274
- 🛠 Xcode 11.5 #273
- 🧹 Purchase cleanup #272
- ♻️ SoftwareMap Protocol #271
- 🕊 Swift 5 #255
  - ⚒️ Xcode 10.2 and macOS 10.14 required to build
- ⬆️ Commandant (0.17.0) #255
- ⬆️ Nimble (8.0.4) #255
- ⬆️ Quick (2.2.0) #255
- ➖ Result #255
- 💡 Update readme with simpler tap usage #241

## [v1.6.4] 🔎 Search Fix - 2020-05-11

- 🐛 Fix search parameters #269

## [v1.6.3] 🗒️ List Fix - 2019-03-31

- 🐛 Fix for list command on macOS 10.14.4 #228
- 🐟 Command completions for fish shell #214  
  thanks, [@ylectric](https://github.com/ylectric)!
- 🎨 SwiftFormat #215
- ⬆️ Nimble (7.3.2) #213
- ⬆️ Quick (1.3.2) #213
- ⬆️ Result (4.1.0) #213

## [v1.6.2] 🗑 Trashy - 2019-01-21

- 🗑🐛 Remove external trash command dependency #210

## [v1.6.1] 🍀 Lucky Again - 2019-01-18

- 🐛 Fix `lucky` command #207
- 🐛 Restore `search` table output #205

## [v1.6.0] 🔗 Links - 2019-01-12

- 👐🏻 Open AppStore.app with just `mas open` #200
- ♻️🌐 Network refactor #198
- 👷🏻‍♀️ Jenkins Pipeline #197
- ✨ New `home`, `open` and `vendor` commands #196
- 🐛📦 Fix paths building installer package #195
- ♻️📚 AppLibrary refactor #193

## [v1.5.0] 🗑 Uninstall - 2018-12-27

- ✨🗑 `uninstall` command #81, #191
- ➕🍺 `trash` formula #191
- 💥 Dropped support for Yosemite and Mavericks #192

## [v1.4.4] 🧹 Cleanup - 2018-12-19

- ⬆️🛠️ Xcode 10.1 #186
- 🐛 Prevent `default.profraw` from getting created (again) #186
- 🚨 Change MasKit dylib version to minor.patch #186
- ✅ Search tests #184

## [v1.4.3] 📦🏛️ MasKit & Carthage - 2018-10-17

Big project restructure to fix issues building for Homebrew.
The `mas` binary now depends on the separate `MasKit.framework`.
Use the `mas.pkg` package to install manually.

- 🧰 MasKit Framework #175
- 📦 Added installer package #178
- 🛠 Xcode 10 & Swift 4.2 #173
- ♻️ Use Carthage rather than CocoaPods to manage dependencies #169  
  thanks, [@tonyarnold](https://github.com/tonyarnold)!
- ⬆️ danger (5.7.0) #176

## [v1.4.2] 🚏📥🙅🏻‍♀️ Sign-In Disabled - 2018-08-12

> Version 1.4.2 was never [released](https://github.com/Homebrew/homebrew-core/pull/31058) through Homebrew.

- 🚏📥🙅🏻‍♀️ Disabled `signin` command on macOS 10.13+ #162
  - An error is immediately returned instead of crashing on newer macOS versions.
  - See #164 for updates on this issue.
- 🐛 Fixed `signout` command #162
- 🐛 Fixed `account` command #165
- ✨ Added price to `search` command when `--price` flag specified #62, #148
- 🎨 `search` output is now formatted in columns #148
- 📺 Added simple bash completion #157
- ➖ CocoaSeeds #155
- ➕🍫 CocoaPods (1.5.3) #155
- 🛠 Xcode 9.4 #153
- 🛠 Xcode 9.3 #141
- 👷🏻‍♀️⚠️ Re-enabled Danger #137

## [v1.4.1] 🚯 Stop Littering - 2018-02-18

- 🐛 Prevent `default.profraw` from getting created #135

## [v1.4.0] 🧟‍♂️ It's Alive! - 2018-02-16

### ✨ Enhancements

- ℹ️ Info command #26, #55
- 🍀 Lucky command #88
- 🕵🏽‍♀️ Include version number in search output #131
- ☝🏻 Support app name(s) arugments for upgrade command #100

### 🔨 Xcode Project

- ⬆️ Swift 4 #114
- ✅ Added test target #127
- 📛 Add build badge to readme #121
- 💎 Configure bundler #118

### 🐙 GitHub

- ⚠️ Added Danger for PR feedback #124
- 📝🤝🎨 Added Contributing Guidelines, Code of Conduct and style guide #117, #126
- 📝 Added CHANGELOG #119
- 📝 Added note about installing through app store first #93

### 🏛 Dependencies #122

- ⬆️ Commandant (0.13.0)
- ⬆️ Result (3.2.4)
- ⬆️ cocoaseeds (0.8.3)
- ⬆️ xcpretty (0.2.8)

### 👷🏻‍♀️ CI Build

- 🍎 Now using [`xcode9.2`](https://docs.travis-ci.com/user/reference/osx#Xcode-9.2) image for `macosx10.13` SDK #120

## [v1.3.1] Better Errors - 2016-09-25

- Descriptive error messages instead of exit codes
- Fixed nullability issue with `list` command
- Simpler upgrade checking

## [v1.3.0] Multiple app install - 2016-09-14

- Fix install of Free apps (#19)
- Install / Upgrade multiple apps at once
- Skip Install if the app is already installed

## [v1.2.2] Secure Password entry - 2016-09-14

- Support reading password from STDIN
- Fix building with Swift 2.3/Xcode 8

## [v1.2.1] - 2016-09-13

- Support reading password from STDIN
- Fix building with Swift 2.3/Xcode 8

## [v1.2.0] Search - 2016-04-16

- `search` command
- Fix `mas list` illegal instruction (#16)

## [v1.1.3] - 2016-02-21

- Fix Illegal Instruction: 4 error (#10)

## [v1.1.2] Upload dSYM correctly - 2016-02-21

- Move the dSYM to the xcarchive

## [v1.1.1] Upload dSYM - 2016-02-21

- Upload dSYM from Travis release

## [v1.1.0] Sign In - 2016-02-13

- Added `signin` command (#3)
- Added `signout` command

## [v1.0.2] Upgrade all - 2015-12-30


Homebrew/mas-tap.rb
class Mas < Formula
  desc "Mac App Store command-line interface"
  homepage "https://github.com/mas-cli/mas"
  url "https://github.com/mas-cli/mas.git",
      tag:      "v1.8.0",
      revision: "9eaf57a5de836ce5e5435a8df14da4aa1b7d7444"
  license "MIT"
  head "https://github.com/mas-cli/mas.git"

  bottle do
    root_url "https://dl.bintray.com/phatblat/mas-bottles"
    sha256 cellar: :any, arm64_big_sur: "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, big_sur:       "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, catalina:      "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, mojave:        "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, high_sierra:   "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, sierra:        "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
    sha256 cellar: :any, el_capitan:    "444796f43ad88de2cc6f8effcde699330867ace7d722b367dd446dd75e1db251"
  end

  depends_on "carthage" => :build
  if Hardware::CPU.arm?
    depends_on xcode: ["12.2", :build]
  else
    depends_on xcode: ["11.4", :build]
  end

  def install
    # Working around build issues in dependencies
    # - Prevent warnings from causing build failures
    # - Prevent linker errors by telling all lib builds to use max size install names
    xcconfig = buildpath/"Overrides.xcconfig"
    xcconfig.write <<~EOS
      GCC_TREAT_WARNINGS_AS_ERRORS = NO
      OTHER_LDFLAGS = -headerpad_max_install_names
    EOS
    ENV["XCODE_XCCONFIG_FILE"] = xcconfig

    # Only build necessary dependencies
    system "carthage", "bootstrap", "--platform", "macOS", "Commandant"
    system "script/install", prefix

    bash_completion.install "contrib/completion/mas-completion.bash" => "mas"
    fish_completion.install "contrib/completion/mas.fish"
  end

  test do
    assert_equal version.to_s, shell_output("#{bin}/mas version").chomp
    assert_include shell_output("#{bin}/mas info 497799835"), "Xcode"
  end
end
